include/jemalloc/internal/thread_event.h

```diff
@@ -0,0 +1,113 @@
+#ifndef JEMALLOC_INTERNAL_THREAD_EVENT_H
+#define JEMALLOC_INTERNAL_THREAD_EVENT_H
+
+#include "jemalloc/internal/tsd.h"
+
+/*
+ * Maximum threshold on thread_allocated_threshold_fast, so that there is no
+ * need to check overflow in malloc fast path. (The allocation size in malloc
+ * fast path never exceeds SC_LOOKUP_MAXCLASS.)
+ */
+#define THREAD_ALLOCATED_THRESHOLD_FAST_MAX				\
+    (UINT64_MAX - SC_LOOKUP_MAXCLASS + 1)
+
+/*
+ * List of all waits, in the following format:
+ *  W(event,		(condition))
+ */
+#define ITERATE_OVER_ALL_WAITS						\
+    W(prof_sample,	(config_prof && opt_prof))
+
+void thread_event_trigger(tsd_t *tsd, uint64_t usize, bool next_time);
+void thread_event_update(tsd_t *tsd);
+
+JEMALLOC_ALWAYS_INLINE void
+thread_event_assert_invariants(tsd_t *tsd) {
+	if (!config_debug) {
+		return;
+	}
+
+	uint64_t thread_allocated = tsd_thread_allocated_get(tsd);
+	uint64_t last_event = tsd_thread_allocated_last_event_get(tsd);
+	uint64_t threshold = tsd_thread_allocated_threshold_get(tsd);
+	uint64_t threshold_fast = tsd_thread_allocated_threshold_fast_get(tsd);
+
+	assert(last_event != threshold);
+	if (last_event < threshold &&
+	    threshold <= THREAD_ALLOCATED_THRESHOLD_FAST_MAX) {
+		assert(threshold_fast == threshold);
+	} else {
+		assert(threshold_fast == 0);
+	}
+
+	/* Both subtractions are intentionally susceptible to underflow. */
+	assert(thread_allocated - last_event < threshold - last_event);
+
+	/*
+	 * The rest of this function checks the wait times, and it can only
+	 * fail (trivially) when both of the following two conditions hold:
+	 * (1) the TSD is just initialized, and
+	 * (2) none of the events has its condition on.
+	 * The thread_allocated == 0 guard below is stronger than needed, but
+	 * having an exactly accurate guard is more complicated to implement.
+	 */
+	if (thread_allocated == 0) {
```

Comment:
I feel from readability perspective, if we put the test early, it would be clearer that the execution of the second half of this function can be completely avoided if the test condition is `true`?

Reply:
Sounds good. Another question, if all conditions are off, why the last assert still hold?

Reply:
The just-initialized `tsd` will (by design) immediately trigger an event (at the first allocation), where the wait time will be set to the max if no condition holds (which is also by design so that we don't need to manually postpone the wait time e.g. in the case when `opt_prof` is off).

Reply:
nit: after consolidating the macros into utility functions, I ended up still following your suggestion since the logic is hidden inside the function call.